onnx/defs/nn/defs.cc

```diff
@@ -2487,4 +2487,187 @@ ONNX_OPERATOR_SET_SCHEMA(
+    }
+  } else {
+    if (ctx.getInputType(2)->tensor_type().shape().dim(0).dim_value() != 2) {
+      fail_shape_inference("tensor contains partial shape for output. It must contain 2 elements");
+    }
+  }
+
+  // validating 'padding' attribute
+  std::vector<int64_t> pads;
+  if (getRepeatedAttribute(ctx, "padding", pads)) {
+    if (pads.size() != 2) {
+      fail_shape_inference("dilation tensor must have 2 dimensions");
+    }
+  }
+
+  // validating 'stride' attribute
+  std::vector<int64_t> strides;
+  if (getRepeatedAttribute(ctx, "stride", strides)) {
+    if (strides.size() != 2) {
+      fail_shape_inference("dilation tensor must have 2 dimensions");
+    }
+  }
+
+  // inferencing final shape
+  auto final_output_shape =
+      ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape();
+
+  auto size_of_output = static_cast<int>(output_shape.size());
+  if (size_of_output > 0) { // try full shape inference
+    // a new dim is always added to the output
+    final_output_shape->add_dim()->set_dim_value(1);
+
+    // input batch dim
+    if (static_cast<size_t>(input_shape.dim_size()) > 2) {
+      *final_output_shape->add_dim() = input_shape.dim(0);
+    }
+
+    // output_shape is appended to the end
+    for (int i = 0; i < size_of_output; ++i) {
+      final_output_shape->add_dim()->set_dim_value(output_shape[i]);
+    }
+  } else { // fallback to rank inference otherwise
+    // a new dim is always added to the output
+    final_output_shape->add_dim();
+
+    // input batch dim
+    if (static_cast<size_t>(input_shape.dim_size()) > 2) {
+      final_output_shape->add_dim();
+    }
+
+    // output_shape is appended to the end
+    int n_output_shape = ctx.getInputType(1)->tensor_type().shape().dim(0).dim_value();
+    for (int i=0; i < n_output_shape; ++i) {
+      final_output_shape->add_dim();
+    }
+  }
+
+  return;
+}
+
+static const char* Col2Im_ver16_doc = R"DOC(
+The operator rearranges the input matrix columns into blocks.
+
+Col2Im behaves like PyTorch's fold for two output spatial dimensions https://pytorch.org/docs/stable/generated/torch.nn.Fold.html
+
+Note that only 4-D output tensors (batched image-like tensors) are supported.
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    Col2Im,
+    16,
+    OpSchema()
+    .Attr(
+        "dilation",
+        "Dilation value along each spatial axis of the kernel. "
+        "`dilation` format should be a 2-tuple as follow (dim_1, dim_2, ...). "
+        "If not present, the dilation defaults to 1 for each spatial axis",
+        AttributeProto::INTS,
+        OPTIONAL_VALUE)
+	  .Attr(
+        "padding",
+        "Padding for the beginning and ending along each spatial axis, it can take any value greater than or equal to 0. "
+        "The value represent the number of pixels added to the beginning and end part of the corresponding axis. "
+        "`padding` format should be a 2-tuple as follow (begin, end). "
```

Comment:
Is it the same padding for every axis? Or, does the user specify begin/end for each axis (which would not be a 2-tuple)?

Reply:
The input is either [H, W] or [N, H, W]. Padding is only meant for the [H, W] dimensions of input.

`padding[0]` means zero padding on both sizes of `input_height` dimension by `padding[0]` pixels
`padding[1]` means zero padding on both sizes of `input_width` dimension by `padding[1]` pixels

so `padding=[0,1]` means 0 pixel on both sides of input height dimension and 1 pixel in both sides of input height dimension

Is that clear?

Reply:
I will add that to the help message

Reply:
I think it would be useful to make attributes like padding uniform for all ops.

Reply:
The same for dilation and stride too?